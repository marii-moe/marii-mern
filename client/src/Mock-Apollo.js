import { buildClientSchema } from 'graphql';
import * as introspectionResult from './server-schema.json';
import { addMockFunctionsToSchema } from 'graphql-tools'
import ApolloClient from 'apollo-client'
import { ApolloLink, Observable } from "apollo-link";
import { InMemoryCache } from "apollo-cache-inmemory";
import { shallow } from 'enzyme'
import { ApolloProvider } from 'react-apollo'
import React from 'react'

const schema = buildClientSchema(introspectionResult);

addMockFunctionsToSchema({schema});

const link = new ApolloLink(operation => {
  return new Observable(observer => {
    const { query, operationName, variables } = operation;
    delay(100)
      .then(() =>
        graphql(schema, print(query), null, null, variables, operationName)
      )
      .then(result => {
        observer.next(result);
        observer.complete();
      })
      .catch(observer.error.bind(observer));
  });
});

function delay(ms) {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, ms);
  });
}

export const mockClient = new ApolloClient({
  link,
  cache: new InMemoryCache()
})

export const apolloRender = (element) => {
  const wrapper = shallow(<ApolloProvider client={mockClient}>{element}</ApolloProvider>                                                                                                     
                          , { context: { client: mockClient }})                                                                                                                                  
  const renderedElement = wrapper.dive().dive().dive()
  renderedElement.root=wrapper
  return renderedElement
}
