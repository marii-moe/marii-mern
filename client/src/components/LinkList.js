import { graphql } from 'react-apollo'
import gql from 'graphql-tag'
import React, { Component } from 'react'
import Link from './Link'
import { LINKS_PER_PAGE } from '../constants'

class LinkList extends Component {
  
  render() {
    if(this.props.allLinksQuery && this.props.allLinksQuery.loading) {
      return <div>Loading</div>
    }
    if(this.props.allLinksQuery && this.props.allLinksQuery.error) {
      return <div>Error</div>
    }
    const isNewPage = this.props.location.pathname.includes('new')
    const linksToRender = this._getLinksToRender(isNewPage)
    const page = parseInt(this.props.match.params.page, 10)
    return (
      <div>
	<div>
	  {linksToRender.map((link, index) => (
	    <Link key={link.id}
		  index={page ? (page-1) * LINKS_PER_PAGE + index : index}
		  link={link}
		  updateStoreAfterVote={this._updateCacheAfterVote}
		  /> 
	  ))}
      </div>
	{isNewPage &&
	 <div className='flex ml4 mv3 gray'>
	 <div className='pointer mr2' onClick={() => this._previousPage()}>Previous</div>
	 <div className='pointer' onClick={() => this._nextPage()}>Next</div>
	 </div>
	}
	</div>
    )
  }
  _getLinksToRender = (isNewPage) => {
    if (isNewPage) {
      return this.props.allLinksQuery.allLinks.links
    }
    const rankedLinks = this.props.allLinksQuery.allLinks.links.slice()
    rankedLinks.sort((l1, l2) => l2.votes.length - l1.votes.length)
    return rankedLinks
  }

  _nextPage = () => {
    const page = parseInt(this.props.match.params.page, 10)
    if(this.props.allLinksQuery.allLinks.pageInfo.hasNextPage) {
      const nextPage = page +1
      this.props.history.push(`/new/${nextPage}`)
    }
  }

  _previousPage = () => {
    const page = parseInt(this.props.match.params.page, 10)
    if(page > 1) {
      const previousPage = page -1
      this.props.history.push(`/new/${previousPage}`)
    }
  }
  
  _updateCacheAfterVote = (store, createVote, linkId) => {
    const isNewPage = this.props.location.pathname.includes('new')
    const page = parseInt(this.props.match.params.page, 10)
    const skip = isNewPage ? (page -1)* LINKS_PER_PAGE : 0
    const first = isNewPage ? LINKS_PER_PAGE : 100
    const data = store.readQuery({ query: ALL_LINKS_QUERY, variables: { first, skip } })
    const votedLink = data.allLinks.links.find(link => link.id === linkId)
    votedLink.votes = createVote.link.votes
    store.writeQuery({ query: ALL_LINKS_QUERY, data })
  }
  _subscribeToNewLinks = () => {
    this.props.allLinksQuery.subscribeToMore({
      document: gql`
        subscription {
          Link(filter: {
            mutation_in: [CREATED]
          }) {
            node {
              id
              url
              description
              postedBy {
                id
                name
              }
              votes {
                id
                user {
                  id
                }
              }
            }
          }
        }
      `,
      updateQuery: (previous, { subscriptionData }) => {
	const newlinks = [
	  subscriptionData.data.Link.node,
	  ...previous.allLinks.links
	]
	console.log(previous)
	const newAllLinks = Object.assign({},previous.allLinks,{allLinks: {links: newlinks}})
	console.log(newAllLinks)
	const result = {
	  ...previous,
	  allLinks: newAllLinks
	}
	return result
      }
    })
  }
  _subscribeToNewVotes = () => {
    this.props.allLinksQuery.subscribeToMore({
      document: gql `
        subscription {
          Vote(filter: {
            mutation_in: [CREATED]
          }) {
            node {
              id
              link {
                id
                url
                description
                postedBy {
                  id
                  name
                }
                votes {
                  id
                  user {
                    id 
                 }
                }
              }
              user {
                id
              }
            }
          }
        }
      `,
      updateQuery: (previous, { subscriptionData }) => {
	const votedLinkIndex = previous.allLinks.links.findIndex(
	  link => link.id === subscriptionData.data.Vote.node.link.id)
	const link = subscriptionData.data.Vote.node.link
	const allLinks = previous.allLinks
	const newLinks = allLinks.links.slice()
	newLinks[votedLinkIndex] = link
	const newAllLinks = Object.assign({}, allLinks, {links: newLinks})
	const result = {
	  ...previous,
	  allLinks: newAllLinks
	}
	return result
      },
    })
  }
  componentDidMount() {
    this._subscribeToNewLinks()
    this._subscribeToNewVotes()
  }

}

export const ALL_LINKS_QUERY = gql`
query AllLinksQuery($first: Int, $skip: Int) {
  allLinks(first: $first, skip: $skip) {
    links {
      id
  #miss    createdAt
      url
      description
      postedBy {
        id
        name
      }
      votes {
        id
        user { 
          id
        }
      }
    }
    pageInfo {
      hasNextPage
    }
  }
}
`

export default graphql(ALL_LINKS_QUERY, {
  name: 'allLinksQuery',
  options: (ownProps) => {
    const page = parseInt(ownProps.match.params.page, 10)
    const isNewPage = ownProps.location.pathname.includes('new')
    const skip = isNewPage ? LINKS_PER_PAGE*(page-1): 0
    const first = isNewPage ? LINKS_PER_PAGE : 100
    return {
      variables: { first, skip }
    }
  }
}) (LinkList)
