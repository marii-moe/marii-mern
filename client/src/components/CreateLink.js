import React, { Component } from 'react'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'
import { GC_USER_ID, LINKS_PER_PAGE } from '../constants'
import { ALL_LINKS_QUERY } from './LinkList'

class CreateLink  extends Component {
  state = {
    description: '',
    url: '',
  }

  render() {
    return (
      <div>
	<div className='flex flex-column mt3'>
	  <input
	    name='description'
	    className='mb2'
	    value={this.state.description}
	    onChange={(event) => this.setState({ description: event.target.value })}
	    type='text'
	    placeholder='A description for the link'
	    />
	    <input
	      name='url'
	      className='mb2'
	      value={this.state.url}
	      onChange={(event) => this.setState({ url: event.target.value })}
	      type='text'
	      placeholder='The URL for the link'
	      />
	</div>
	<button onClick={() => this._createLink()} >
	  Submit
	</button>
      </div>
    )
  }

  _createLink = async () => {
    const postedById = localStorage.getItem(GC_USER_ID)
    const { description, url } = this.state
    await this.props.createLinkMutation({
      variables: {
	description,
	url,
      },
      update: this._updateStoreWithLink
    })
    this.props.history.push(`/new/1`)
  }

  _updateStoreWithLink = (store, { data: {createLink } }) => {
    const first = LINKS_PER_PAGE
    const skip = 0
    const data = store.readQuery({
      query: ALL_LINKS_QUERY,
      variables: { first, skip}
    })
    data.allLinks.links.splice(0,0,createLink)
    if(data.allLinks.links.length > first)
      data.allLinks.links.pop()
    store.writeQuery({
      query: ALL_LINKS_QUERY,
      data,
      variables: { first, skip, }
    })
  }
}

export { CreateLink as _CreateLink }

const CREATE_LINK_MUTATION = gql`
mutation CreateLinkMutation($description: String!, $url: String!) {
  createLink(
    description: $description,
    url: $url
  ) {
    id
    #missing createdAt
    url
    description
    postedBy {
      id
      name
    }
  }
}
`

export default graphql(CREATE_LINK_MUTATION, { name: 'createLinkMutation' })(CreateLink)
