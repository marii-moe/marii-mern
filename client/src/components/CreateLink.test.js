import { ApolloProvider } from 'react-apollo'
import React from 'react';
import ReactDOM from 'react-dom';
import { _CreateLink }  from './CreateLink'
import CreateLink  from './CreateLink'
import { mockClient, apolloRender } from "../Mock-Apollo"
import { shallow } from 'enzyme'
import { LINKS_PER_PAGE } from '../constants'
import { ALL_LINKS_QUERY } from './LinkList'

it('renders successfully', async () => {
  localStorage.getItem = () => { return 1 }
  const div = document.createElement('div')
  const wrapper = apolloRender(<CreateLink/>)
  expect(wrapper).toMatchSnapshot()
})

it('redirects back to LinkList', async () => {
  const createLink = new _CreateLink()
  const createLinkMutationMock = () => { return true }
  const historyMock = []
  createLink.props = { client: mockClient,
		       history: historyMock,
		       createLinkMutation: createLinkMutationMock
		     }
  await createLink._createLink()
  expect(historyMock[0]).toEqual('/new/1')
})

it('updates the store', async () => {
  const createLink = new _CreateLink()
  const readQuery = jest.fn()
  const writeQuery = jest.fn()
  readQuery.mockReturnValue({allLinks: {links: []}})
  const store = { readQuery: readQuery, writeQuery: writeQuery }
  
  const links = {data: {createLink: { description: 'Mock link'} }}
  createLink._updateStoreWithLink( store , links)
  const writtenToStore = store.writeQuery.mock.calls[0][0].data.allLinks.links[0].description
  expect(writtenToStore).toEqual('Mock link')
})

it('submits a link', async () => {
  localStorage.getItem = () => { return 1 }
  const div = document.createElement('div')
  const wrapper = apolloRender(<CreateLink/>)
  const submitfunction = jest.fn()
  wrapper.instance()._createLink = submitfunction
  wrapper.find('button').simulate('click');
  expect(submitfunction.mock.calls.length).toEqual(1)
})

it('updates state when textboxes are updated', async () => {
  localStorage.getItem = () => { return 1 }
  const div = document.createElement('div')
  const wrapper = apolloRender(<CreateLink/>)
  wrapper.find('input').at(0).simulate('change', {target: {value: 'My new description'}});
  expect(wrapper.instance().state.description).toEqual('My new description')
  wrapper.find('input').at(1).simulate('change', {target: {value: 'My new link'}});
  expect(wrapper.instance().state.url).toEqual('My new link')
})

it('updates the store and removes link if over 5', async () => {
  const createLink = new _CreateLink()
  const readQuery = jest.fn()
  const writeQuery = jest.fn()
  readQuery.mockReturnValue({allLinks: {links: [{},{},{},{},{}]}})
  const store = { readQuery: readQuery, writeQuery: writeQuery }
  
  const links = {data: {createLink: { description: 'Mock link'} }}
  createLink._updateStoreWithLink( store , links)
  const writtenLinks = store.writeQuery.mock.calls[0][0].data.allLinks.links
  const writtenToStore = writtenLinks[0].description
  expect(writtenToStore).toEqual('Mock link')
  expect(writtenLinks.length).toEqual(5)
})

