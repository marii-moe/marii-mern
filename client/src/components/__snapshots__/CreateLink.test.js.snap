// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render successfully 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CreateLink
    createLinkMutation={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="flex flex-column mt3"
>
          <input
                    className="mb2"
                    name="description"
                    onChange={[Function]}
                    placeholder="A description for the link"
                    type="text"
                    value=""
          />
          <input
                    className="mb2"
                    name="url"
                    onChange={[Function]}
                    placeholder="The URL for the link"
                    type="text"
                    value=""
          />
</div>,
        <button
          onClick={[Function]}
>
          Submit
</button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="mb2"
              name="description"
              onChange={[Function]}
              placeholder="A description for the link"
              type="text"
              value=""
/>,
            <input
              className="mb2"
              name="url"
              onChange={[Function]}
              placeholder="The URL for the link"
              type="text"
              value=""
/>,
          ],
          "className": "flex flex-column mt3",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "mb2",
              "name": "description",
              "onChange": [Function],
              "placeholder": "A description for the link",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "mb2",
              "name": "url",
              "onChange": [Function],
              "placeholder": "The URL for the link",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Submit",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Submit",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="flex flex-column mt3"
>
            <input
                        className="mb2"
                        name="description"
                        onChange={[Function]}
                        placeholder="A description for the link"
                        type="text"
                        value=""
            />
            <input
                        className="mb2"
                        name="url"
                        onChange={[Function]}
                        placeholder="The URL for the link"
                        type="text"
                        value=""
            />
</div>,
          <button
            onClick={[Function]}
>
            Submit
</button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="mb2"
                name="description"
                onChange={[Function]}
                placeholder="A description for the link"
                type="text"
                value=""
/>,
              <input
                className="mb2"
                name="url"
                onChange={[Function]}
                placeholder="The URL for the link"
                type="text"
                value=""
/>,
            ],
            "className": "flex flex-column mt3",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "mb2",
                "name": "description",
                "onChange": [Function],
                "placeholder": "A description for the link",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "mb2",
                "name": "url",
                "onChange": [Function],
                "placeholder": "The URL for the link",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Submit",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Submit",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "client": ApolloClient {
        "cache": InMemoryCache {
          "addTypename": true,
          "config": Object {
            "addTypename": true,
            "dataIdFromObject": [Function],
            "fragmentMatcher": HeuristicFragmentMatcher {},
            "storeFactory": [Function],
          },
          "data": ObjectCache {
            "data": Object {},
          },
          "optimistic": Array [],
          "silenceBroadcast": false,
          "watches": Array [],
        },
        "defaultOptions": Object {},
        "disableNetworkFetches": false,
        "link": ApolloLink {
          "request": [Function],
        },
        "mutate": [Function],
        "query": [Function],
        "queryDeduplication": true,
        "reFetchObservableQueries": [Function],
        "resetStore": [Function],
        "ssrMode": false,
        "store": DataStore {
          "cache": InMemoryCache {
            "addTypename": true,
            "config": Object {
              "addTypename": true,
              "dataIdFromObject": [Function],
              "fragmentMatcher": HeuristicFragmentMatcher {},
              "storeFactory": [Function],
            },
            "data": ObjectCache {
              "data": Object {},
            },
            "optimistic": Array [],
            "silenceBroadcast": false,
            "watches": Array [],
          },
        },
        "version": "2.0.4",
        "watchQuery": [Function],
      },
    },
  },
}
`;

exports[`renders successfully 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CreateLink
    createLinkMutation={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="flex flex-column mt3"
>
          <input
                    className="mb2"
                    name="description"
                    onChange={[Function]}
                    placeholder="A description for the link"
                    type="text"
                    value=""
          />
          <input
                    className="mb2"
                    name="url"
                    onChange={[Function]}
                    placeholder="The URL for the link"
                    type="text"
                    value=""
          />
</div>,
        <button
          onClick={[Function]}
>
          Submit
</button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="mb2"
              name="description"
              onChange={[Function]}
              placeholder="A description for the link"
              type="text"
              value=""
/>,
            <input
              className="mb2"
              name="url"
              onChange={[Function]}
              placeholder="The URL for the link"
              type="text"
              value=""
/>,
          ],
          "className": "flex flex-column mt3",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "mb2",
              "name": "description",
              "onChange": [Function],
              "placeholder": "A description for the link",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "mb2",
              "name": "url",
              "onChange": [Function],
              "placeholder": "The URL for the link",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Submit",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Submit",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="flex flex-column mt3"
>
            <input
                        className="mb2"
                        name="description"
                        onChange={[Function]}
                        placeholder="A description for the link"
                        type="text"
                        value=""
            />
            <input
                        className="mb2"
                        name="url"
                        onChange={[Function]}
                        placeholder="The URL for the link"
                        type="text"
                        value=""
            />
</div>,
          <button
            onClick={[Function]}
>
            Submit
</button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="mb2"
                name="description"
                onChange={[Function]}
                placeholder="A description for the link"
                type="text"
                value=""
/>,
              <input
                className="mb2"
                name="url"
                onChange={[Function]}
                placeholder="The URL for the link"
                type="text"
                value=""
/>,
            ],
            "className": "flex flex-column mt3",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "mb2",
                "name": "description",
                "onChange": [Function],
                "placeholder": "A description for the link",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "mb2",
                "name": "url",
                "onChange": [Function],
                "placeholder": "The URL for the link",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Submit",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Submit",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "client": ApolloClient {
        "cache": InMemoryCache {
          "addTypename": true,
          "config": Object {
            "addTypename": true,
            "dataIdFromObject": [Function],
            "fragmentMatcher": HeuristicFragmentMatcher {},
            "storeFactory": [Function],
          },
          "data": ObjectCache {
            "data": Object {},
          },
          "optimistic": Array [],
          "silenceBroadcast": false,
          "watches": Array [],
        },
        "defaultOptions": Object {},
        "disableNetworkFetches": false,
        "link": ApolloLink {
          "request": [Function],
        },
        "mutate": [Function],
        "query": [Function],
        "queryDeduplication": true,
        "reFetchObservableQueries": [Function],
        "resetStore": [Function],
        "ssrMode": false,
        "store": DataStore {
          "cache": InMemoryCache {
            "addTypename": true,
            "config": Object {
              "addTypename": true,
              "dataIdFromObject": [Function],
              "fragmentMatcher": HeuristicFragmentMatcher {},
              "storeFactory": [Function],
            },
            "data": ObjectCache {
              "data": Object {},
            },
            "optimistic": Array [],
            "silenceBroadcast": false,
            "watches": Array [],
          },
        },
        "version": "2.0.4",
        "watchQuery": [Function],
      },
    },
  },
}
`;
